src/app/CommandSender.cpp

```diff
@@ -70,27 +72,22 @@ void CommandSender::OnMessageReceived(Messaging::ExchangeContext * apExchangeCon
                                       const PayloadHeader & aPayloadHeader, System::PacketBufferHandle aPayload)
 {
     CHIP_ERROR err = CHIP_NO_ERROR;
-    // Assert that the exchange context matches the client's current context.
-    // This should never fail because even if SendCommandRequest is called
-    // back-to-back, the second call will call Close() on the first exchange,
-    // which clears the OnMessageReceived callback.
 
-    VerifyOrDie(apExchangeContext == mpExchangeCtx);
-
-    // Verify that the message is an Invoke Command Response.
-    // If not, close the exchange and free the payload.
-    if (!aPayloadHeader.HasMessageType(Protocols::InteractionModel::MsgType::InvokeCommandResponse))
-    {
-        apExchangeContext->Close();
-        mpExchangeCtx = nullptr;
-        goto exit;
-    }
+    VerifyOrExit(apExchangeContext == mpExchangeCtx, err = CHIP_ERROR_INCORRECT_STATE);
+    VerifyOrExit(aPayloadHeader.HasMessageType(Protocols::InteractionModel::MsgType::InvokeCommandResponse),
+                 err = CHIP_ERROR_INVALID_MESSAGE_TYPE);
 
     err = ProcessCommandMessage(std::move(aPayload), CommandRoleId::SenderId);
     SuccessOrExit(err);
 
 exit:
+    ChipLogFunctError(err);
+
+    // We shall close the exchange instead of abort to allow the ExchangeMgr complete ack if there is any.
```

Comment:
```suggestion
    // Close the exchange cleanly so that the ExchangeManager will send an ack for the message we just received.
    // This needs to be done before the Reset() call, because Reset() aborts mpExchangeCtx if its not null.
```